On Premises example AWS infrastructure

cmwd opened the pull request.

morganfree commented on aws/ecs/README.md:
```diff
@@ -0,0 +1,51 @@
+# CKEditor Collaboration Server On-Premise AWS ECS example
```
There should more information about network, its subnets, etc. IMO.

morganfree commented on aws/ecs/lib/stack-config.ts:
```diff
@@ -0,0 +1,32 @@
+			Annotations.of( this ).addError( 'Context parameter "dockerRegistryAuthToken" is required' );
+		}
+
+		this.version = this.node.tryGetContext( 'version' ) ?? '3.7.1';
```
`3.8.0` is the latest but I think that the `latest` tag should work on our docker registry as well.

morganfree reviewed (changes requested):
I did not check the infrastructure - I would like to run the deploy but I could not do it. I do not want to read more about cdk because it will be a good test for the documentation for a non experienced client.

1. There should be a precise list of commands which leads to the deploy
2. There should be list of global dependencies/requirements
3. I guess there will be more examples for the same infrastructure but with a different tools, e.g. terraform? This is related to directory structure I guess.

cmwd commented on aws/ecs/package.json:
```diff
@@ -0,0 +1,41 @@
+  "scripts": {
+    "build": "tsc",
+    "watch": "tsc -w",
+    "test": "jest",
```
No, I'll remove it

cmwd commented on aws/ecs/package.json:
```diff
@@ -0,0 +1,41 @@
+    "@types/node": "10.17.27",
+    "aws-cdk": "1.76.0",
+    "eslint": "^6.8.0",
+    "jest": "^26.4.2",
```
☝️

cmwd commented on aws/ecs/package.json:
```diff
@@ -0,0 +1,41 @@
+    "@cksource-cs/eslint-config-cs-module": "^2.1.1",
+    "@types/jest": "^26.0.19",
+    "@types/node": "10.17.27",
+    "aws-cdk": "1.76.0",
```
This is autogenerated and probably could be removed. But I'd prefer to keep it and use it instead of global dependencies. I like it this way because it's a more stable approach.

czerwonkabartosz commented on .gitignore:
```diff
@@ -0,0 +1,10 @@
+node_modules
```
Repository name should be changed `ckeditor-cs-on-premise-infrastructure-examples` -> `ckeditor-cs-on-premises-infrastructure-examples`.

Maybe you can also remove `examples` from the name.

czerwonkabartosz commented on aws/ecs/README.md:
```diff
@@ -0,0 +1,73 @@
+### Database
+Databases are running in an isolated network (no in/out internet access) with backup and encryption enabled.
+MySQL (RDS Aurora) is running on R5 large instances.
+Redis (ElastiCache cluster) is running on T3 large instances.
```
T3? Maybe we should recommend M type?

czerwonkabartosz commented on aws/ecs/README.md:
```diff
@@ -0,0 +1,73 @@
+Redis (ElastiCache cluster) is running on T3 large instances.
+
+### S3
+Private encrypted bucket for editor files.
```
IMO not only editor files - EI files too.

czerwonkabartosz commented on aws/ecs/README.md:
```diff
@@ -0,0 +1,73 @@
+- ElastiCache Redis in cluster mode - medium T3 instance
+- S3 bucket
+
+To provide high-security standards we keep infrastructure resources in private/isolated subnets whenever it's possible. The data is encrypted in the bucket, as well as in the database and Redis.
```
I would love to see some nice and beautiful graph with this infrastructure. 

https://www.google.com/search?q=aws+infrastructure&source=lnms&tbm=isch&sa=X&ved=2ahUKEwi1wtvP6oHuAhVOCRAIHejNC1UQ_AUoAXoECAEQAw&biw=1920&bih=944

czerwonkabartosz commented on aws/ecs/lib/ck-editor-cs-on-premise-infrastructure-stack.ts:
```diff
@@ -0,0 +1,80 @@
+
+		Tags.of( this ).add(
+			'Application',
+			'CKEditor - Cloud Server On-Premises Installation'
```
`Cloud` -> `Collaboration`

czerwonkabartosz commented on aws/ecs/lib/cs-environment-config.ts:
```diff
@@ -0,0 +1,42 @@
+	[ 'APPLICATION_HTTP_PORT', '80' ]
+];
+
+export class CsEnvironmentConfig extends Construct {
```
IMO `CsE...` -> `CSE...`

czerwonkabartosz commented on aws/ecs/lib/resource/application.ts:
```diff
@@ -0,0 +1,129 @@
+
+		this.taskDefinition = new Ec2TaskDefinition( this, 'TaskDefinition' );
+
+		this.containerDefinition = this.taskDefinition.addContainer( 'CloudServicesContainer', {
```
`CloudServicesContainer` -> `CollaborationServerContainer`

czerwonkabartosz commented on aws/ecs/lib/resource/application.ts:
```diff
@@ -0,0 +1,129 @@
+				DATABASE_DATABASE: props.database.name,
+				STORAGE_DRIVER: 's3',
+				STORAGE_ENDPOINT: props.storage.bucket.bucketDomainName,
+				STORAGE_BUCKET: props.storage.bucket.bucketName,
```
IMO you can set `COLLABORATION_STORAGE_DRIVER` too.

https://ckeditor.com/docs/cs/latest/onpremises/cs-onpremises/installation/docker.html

```
STORAGE_DRIVER                           - optional (default: "filesystem")
STORAGE_LOCATION                         - optional (default: "/var/cs/easyimage", required for file system driver)
STORAGE_REGION                           - optional (required for AWS S3 driver)
STORAGE_ACCESS_KEY_ID                    - optional (required for AWS S3 driver)
STORAGE_SECRET_ACCESS_KEY                - optional (required for AWS S3 driver)
STORAGE_BUCKET                           - optional (required for AWS S3 driver)
STORAGE_ENDPOINT                         - optional (required for AWS S3 driver)
STORAGE_ACCOUNT_NAME                     - optional (required for Azure Blob Storage driver)
STORAGE_ACCOUNT_KEY                      - optional (required for Azure Blob Storage driver)
STORAGE_CONTAINER                        - optional (required for Azure Blob Storage driver)
COLLABORATION_STORAGE_DRIVER             - optional (default: "database")
COLLABORATION_STORAGE_LOCATION           - optional (default: "/var/cs/easyimage", required for file system driver)
COLLABORATION_STORAGE_ACCESS_KEY_ID      - optional (required for AWS S3 driver)
COLLABORATION_STORAGE_SECRET_ACCESS_KEY  - optional (required for AWS S3 driver)
COLLABORATION_STORAGE_REGION             - optional (required for AWS S3 driver)
COLLABORATION_STORAGE_ENDPOINT           - optional (required for AWS S3 driver)
COLLABORATION_STORAGE_BUCKET             - optional (required for AWS S3 driver)
COLLABORATION_STORAGE_ACCOUNT_NAME       - optional (required for Azure Blob Storage driver)
COLLABORATION_STORAGE_ACCOUNT_KEY        - optional (required for Azure Blob Storage driver)
COLLABORATION_STORAGE_CONTAINER          - optional (required for Azure Blob Storage driver)
```

czerwonkabartosz commented on aws/ecs/lib/resource/application.ts:
```diff
@@ -0,0 +1,129 @@
+			);
+
+			this.loadBalancedEc2Service.service.addPlacementStrategies(
+				PlacementStrategy.spreadAcrossInstances(),
```
Please add strategy for spread AZ. 

AZ
Instances
CPU

czerwonkabartosz commented on aws/ecs/lib/resource/application.ts:
```diff
@@ -0,0 +1,129 @@
+
+		this.targetGroup.configureHealthCheck( { path: '/health' } );
+
+		this.cluster.connections.allowFrom( this.loadBalancer, Port.allTraffic() );
```
Hmmm all traffic is needed? IMO you can allow only specific ports.

czerwonkabartosz commented on aws/ecs/lib/resource/redis.ts:
```diff
@@ -0,0 +1,59 @@
+export class Redis extends Construct {
+	public static readonly PORT: number = 6379;
+
+	public static readonly CLUSTER_BUS_PORT: number = 16379
```
🤔  What is it?

czerwonkabartosz commented on aws/ecs/lib/resource/redis.ts:
```diff
@@ -0,0 +1,59 @@
+			engineVersion: '6.x',
+			cacheNodeType: 'cache.t3.medium',
+			replicasPerNodeGroup: 1,
+			numNodeGroups: 3,
```
IMO 2 will be enough.

czerwonkabartosz commented on aws/ecs/lib/resource/redis.ts:
```diff
@@ -0,0 +1,59 @@
+			numNodeGroups: 3,
+			automaticFailoverEnabled: true,
+			autoMinorVersionUpgrade: true,
+			replicationGroupDescription: 'CKEditor redis cluster',
```
Add `Collaboration Server`

czerwonkabartosz commented on README.md:
```diff
@@ -0,0 +1,6 @@
+This repository contains a set of easy to use scripts that helps to provision recommended infrastructure for CKEditor Collaboration Server On-Premises. You can use the script as a reference or set up a production-ready service.
+
+[Read more about Collaboration Server](https://ckeditor.com/docs/cs/latest/onpremises/cs-onpremises/overview.html)
+
```
Could you add here some general and generic overview of the architecture. A few words that LB is needed, etc. And add link to https://ckeditor.com/docs/cs/latest/onpremises/cs-onpremises/architecture.html . 

And in AWS ECS you can use more specific definitions for AWS.

czerwonkabartosz replied:
https://www.notion.so/Common-tickets-answers-d3fc855ba4924230a2f684403f145698#900863acd4c14967abb2b6b5eb6ecb29

czerwonkabartosz commented on aws/ecs/README.md:
```diff
@@ -0,0 +1,73 @@
+**Note**:
+For easier work with CDK, we recommend using it together with AWS credentials manager like [aws-vault](https://github.com/99designs/aws-vault).
+
+## Parameters
```
It should be before `Quick start`

czerwonkabartosz commented on aws/ecs/README.md:
```diff
@@ -0,0 +1,73 @@
+[Read more about AWS CDK context parameters](https://docs.aws.amazon.com/cdk/latest/guide/context.html)
+
+- `dockerRegistryAuthToken` - Authentication token to `docker.cke-cs.com` repository
+- `version` - Collaboration Server version (default 3.7.1)
```
How to change it and use a different version?

czerwonkabartosz commented on aws/ecs/README.md:
```diff
@@ -0,0 +1,73 @@
+
+- `dockerRegistryAuthToken` - Authentication token to `docker.cke-cs.com` repository
+- `version` - Collaboration Server version (default 3.7.1)
+- `env` - A comma-separated list of CKEditor Collaboration Server configuration ([see available options](https://ckeditor.com/docs/cs/latest/onpremises/cs-onpremises/installation/docker.html#docker-container-environment-variables))
```
☝️ Some examples will be useful.

czerwonkabartosz commented on aws/ecs/bin/ckeditor-cs-on-premise-infrastructure.ts:
```diff
@@ -0,0 +1,14 @@
+const app: App = new App();
+
+// eslint-disable-next-line no-new
+new CkEditorCsOnPremiseStack( app, 'CKEditorCloudServicesOnPremisesInstallation' );
```
Try to unify names.

czerwonkabartosz reviewed (changes requested):
Generally, it looks nice 😍  Good job 👍 

I would love to run it but I know CDK and it will be easy for me for sure. So, maybe @dawidpawlowski could test it before me?

dawidpawlowski commented:
> I would love to run it but I know CDK and it will be easy for me for sure. So, maybe @dawidpawlowski could test it before me?

I don't know too, but I can try yo test it 😛

czerwonkabartosz reviewed (approved):
For me, it looks good 👍  

We have to wait for @dawidpawlowski test and review.

dawidpawlowski commented:
# Tests results:

## General feelings
- The user experience is very high, you need only 3-4 commands to execute and the entire On-Premises environment is ready.
- Installation is very simple so it doesn't require any CDK knowledge at all. For the user - great, for me - I am disappointed 😄 
- Management panel is working properly.

### To fix: 
- Missing `deploy`command here: https://github.com/cksource/ckeditor-cs-on-premises-infrastructure/pull/1/files#diff-c5bdb42fe03c216a7e842173db94b7de496022c6ce57d1e12aaff3796220ca36R32
- Is it possible to show:
```
CKEditorServerEndpoint=http://CKEdi-Appli-LWDNGY6G8YLX-2142793784.eu-west-3.elb.amazonaws.com
```
instead of
```
CKEditorCloudServicesOnPremisesInstallation.ApplicationLoadBalancedEc2ServiceServiceURLCCD599E7 = http://CKEdi-Appli-LWDNGY6G8YLX-2142793784.eu-west-3.elb.amazonaws.com
```
at the end of the installation?
- Only one On-Premises service instance is run. IMO should be at least 2 in the separate AZ.
- It would be useful to know how to change the configuration of the environment variables (i.e. run the deploy command again with the changes)
- 2 x E2E tests related to documents storage fails because environment variables: `STORAGE_ACCESS_KEY_ID` and `STORAGE_SECRET_ACCESS_KEY` are not set. If this cannot be done automatically, you must add a note to the command that these variables must be set.
- Even after setting these variables manually, 2 x tests still failing for document storage. In the logs, I can see: `Access Denied`. Probably there are no proper permissions for S3. Log details:
```
{"level":60,"time":"2021-01-15T14:50:07.408Z","msg":"Access Denied","data":{"name":"Error","message":"Access Denied","stack":"AccessDenied: Access Denied\n at Request.extractError (/usr/src/app/node_modules/.pnpm/registry.npmjs.org/aws-sdk/2.523.0/node_modules/aws-sdk/lib/services/s3.js:585:35)\n at Request.callListeners (/usr/src/app/node_modules/.pnpm/registry.npmjs.org/aws-sdk/2.523.0/node_modules/aws-sdk/lib/sequential_executor.js:106:20)\n at Request.emit (/usr/src/app/node_modules/.pnpm/registry.npmjs.org/aws-sdk/2.523.0/node_modules/aws-sdk/lib/sequential_executor.js:78:10)\n at Request.emit (/usr/src/app/node_modules/.pnpm/registry.npmjs.org/aws-sdk/2.523.0/node_modules/aws-sdk/lib/request.js:683:14)\n at Request.transition (/usr/src/app/node_modules/.pnpm/registry.npmjs.org/aws-sdk/2.523.0/node_modules/aws-sdk/lib/request.js:22:10)\n at AcceptorStateMachine.runTo (/usr/src/app/node_modules/.pnpm/registry.npmjs.org/aws-sdk/2.523.0/node_modules/aws-sdk/lib/state_machine.js:14:12)\n at /usr/src/app/node_modules/.pnpm/registry.npmjs.org/aws-sdk/2.523.0/node_modules/aws-sdk/lib/state_machine.js:26:10\n at Request.<anonymous> (/usr/src/app/node_modules/.pnpm/registry.npmjs.org/aws-sdk/2.523.0/node_modules/aws-sdk/lib/request.js:38:9)\n at Request.<anonymous> (/usr/src/app/node_modules/.pnpm/registry.npmjs.org/aws-sdk/2.523.0/node_modules/aws-sdk/lib/request.js:685:12)\n at Request.callListeners (/usr/src/app/node_modules/.pnpm/registry.npmjs.org/aws-sdk/2.523.0/node_modules/aws-sdk/lib/sequential_executor.js:116:18)\n at Request.emit (/usr/src/app/node_modules/.pnpm/registry.npmjs.org/aws-sdk/2.523.0/node_modules/aws-sdk/lib/sequential_executor.js:78:10)\n at Request.emit (/usr/src/app/node_modules/.pnpm/registry.npmjs.org/aws-sdk/2.523.0/node_modules/aws-sdk/lib/request.js:683:14)\n at Request.transition (/usr/src/app/node_modules/.pnpm/registry.npmjs.org/aws-sdk/2.523.0/node_modules/aws-sdk/lib/request.js:22:10)\n at AcceptorStateMachine.runTo (/usr/src/app/node_modules/.pnpm/registry.npmjs.org/aws-sdk/2.523.0/node_modules/aws-sdk/lib/state_machine.js:14:12)\n at /usr/src/app/node_modules/.pnpm/registry.npmjs.org/aws-sdk/2.523.0/node_modules/aws-sdk/lib/state_machine.js:26:10\n at Request.<anonymous> (/usr/src/app/node_modules/.pnpm/registry.npmjs.org/aws-sdk/2.523.0/node_modules/aws-sdk/lib/request.js:38:9)\n at Request.<anonymous> (/usr/src/app/node_modules/.pnpm/registry.npmjs.org/aws-sdk/2.523.0/node_modules/aws-sdk/lib/request.js:685:12)\n at Request.callListeners (/usr/src/app/node_modules/.pnpm/registry.npmjs.org/aws-sdk/2.523.0/node_modules/aws-sdk/lib/sequential_executor.js:116:18)\n at callNextListener (/usr/src/app/node_modules/.pnpm/registry.npmjs.org/aws-sdk/2.523.0/node_modules/aws-sdk/lib/sequential_executor.js:96:12)\n at IncomingMessage.onEnd (/usr/src/app/node_modules/.pnpm/registry.npmjs.org/aws-sdk/2.523.0/node_modules/aws-sdk/lib/event_listeners.js:307:13)\n at IncomingMessage.emit (events.js:323:22)\n at IncomingMessage.EventEmitter.emit (domain.js:482:12)\n at endReadableNT (_stream_readable.js:1204:12)\n at processTicksAndRejections (internal/process/task_queues.js:84:21)","message":"Access Denied","code":"AccessDenied","region":null,"time":"2021-01-15T14:50:07.408Z","requestId":"54B8654DD35710C7","extendedRequestId":"+uAYWiifZZe2G4BTzXzvYwr1dKmKuvi+ywuncCskAdhgkATjYEmjPWCLg4J1nCprHsfkWuvFaJo=","statusCode":403,"retryable":false,"retryDelay":155.79868697357594}}
```

dawidpawlowski reviewed (changes requested):
Please fix things from my previous comment. Especially the case with S3.

czerwonkabartosz commented:
Please remember about it

SlawekGreczyn commented on README.md:
```diff
@@ -0,0 +1,11 @@
+This repository contains a set of easy to use scripts that helps to provision recommended infrastructure for CKEditor Collaboration Server On-Premises. You can use the script as a reference or set up a production-ready service.
+
+Regardless of the infrastructure provider, we're recommending to use layered architecture for the separation of data from the application layer.
```
I would add more information about the `layered architecture`, it can be not too clean for a customers.

cmwd replied:
I've added a link to: https://ckeditor.com/docs/cs/latest/onpremises/cs-onpremises/architecture.html#overview

SlawekGreczyn commented on aws/ecs/README.md:
```diff
@@ -0,0 +1,74 @@
+
+To use AWS CDK first You need an AWS account and AWS credentials. Read more about how to configure the environment [on official CDK docs](https://docs.aws.amazon.com/cdk/latest/guide/work-with.html).
+
+- install dependencies
```
I would add that the commands should be executed on a developer's local computer

cmwd replied:
Hmm, this is not always true. The script can be run on CI

SlawekGreczyn commented on aws/ecs/README.md:
````diff
@@ -0,0 +1,74 @@
+```
+
+**Note**:
+For easier work with CDK, we recommend using it together with AWS credentials manager like [aws-vault](https://github.com/99designs/aws-vault).
````
Explain why it is recommended?

SlawekGreczyn commented on aws/ecs/README.md:
```diff
@@ -0,0 +1,74 @@
+- private subnet for application cluster
+- isolated subnet for databases
+
+### Autoscaling Group
```
Am i right that **`Autoscaling Group`** occurs twice?

cmwd replied:
Fixed

SlawekGreczyn commented on aws/ecs/lib/resource/database.ts:
```diff
@@ -0,0 +1,76 @@
+  network: Network;
+}
+
+export class Database extends Construct {
```
I would add a piece of information, about what each file does, ex:   
  
// The script prepares databases infrastructure.   
// It will store all the persistent data from the CKEditor like comments, suggestions etc.  
// ....  
export class Database extends Construct {

cmwd commented:
> Only one On-Premises service instance is run. IMO should be at least 2 in the separate AZ.

Fixed, I've set desired number of instances to 2

> Command to destroy the whole environment will be useful.

Added

> 2 x E2E tests related to documents storage fails because environment variables: STORAGE_ACCESS_KEY_ID and STORAGE_SECRET_ACCESS_KEY are not set. If this cannot be done automatically, you must add a note to the command that these variables must be set.
Even after setting these variables manually, 2 x tests still failing for document storage. In the logs, I can see: Access Denied. Probably there are no proper permissions for S3. Log details:

This error happen due to S3 bucket name provisioning lag. Switching to regional domain address fixed the problem. More info here: https://aws.amazon.com/premiumsupport/knowledge-center/s3-http-307-response/

Setting AWS credentials variables is not needed in the environment managed by AWS service.

dawidpawlowski reviewed (approved):
I confirm. The script is working properly now. 👍  👏

dawidpawlowski commented:
Hmmm, is it not possible to skip bucket destroy and continue destroying other stuff?

![image](https://user-images.githubusercontent.com/2760195/105484692-08ff2000-5cac-11eb-8d6f-00c5dd0892d7.png)

cmwd commented:
@dawidpawlowski what has happened now is that he removed whole stack (you can confirm that in CloudFormation stack, event tab) except S3 because You have some files inside. You can go to S3 console, empty the bucket and try to run cdk destroy again, or do it manually in CloudFormation console.
